File: js/sorts/quick_sort.js
global.stage = 0;

let array = [9, 12, 2, 3, 5, 2, 5, 6, 4, 3, 7, 10, 1, 12, 8, 11];
/**
 * @param {Array} array
 */
function quickSort(array) {
	if (array.length == 0) return [];
	let left = [],
		right = [],
		// Take the pivot as the first element of array
		pivot = array[0];

	for (let i = 1; i < array.length; i++) {
		if (array[i] < pivot) left.push(array[i]);
		else right.push(array[i]);
	}
	console.log(`stage ${stage}:`, left, pivot, right);
	++stage;
	return quickSort(left).concat(pivot, quickSort(right));
}

console.log(quickSort(array.slice()));

File: python/sorts/quick_sort.py
def quick_sort(a: list):
    n = len(a)
    if n <= 1:
        return

    i = -1
    j = 0
    pivot = a[n - 1]
    
    while j < n - 1:
        if a[j] <= pivot:
            i += 1
            a[j], a[i] = a[i], a[j]
        j += 1

    pivot_i = i + 1

    a[pivot_i], a[j] = a[j], a[pivot_i]

    left = a[:pivot_i]
    right = a[pivot_i + 1:]

    quick_sort(left)
    quick_sort(right)

    length_l = len(left)
    length_r = len(right)

    for k in range(length_l):
        a[k] = left[k]
    for k in range(length_r):
        a[length_l + k + 1] = right[k]
